src/nifreeze/data/dmri/utils.py

```diff
@@ -0,0 +1,317 @@
+        ...     ]
+        ... )
+        array([[   1,    0,    0,    0],
+               [   0,    1,    0, 1000],
+               [   0,    0,    1, 2000],
+               [   0,    0,    0,    0],
+               [   0,    0,    0,    0]], dtype=int16)
+
+    Column-major inputs are automatically transposed when an expected
+    number of diffusion volumes is provided::
+
+        >>> format_gradients(
+        ...     [[1, 0], [0, 1], [0, 0], [1000, 2000]],
+        ... )
+        array([[   1,    0,    0, 1000],
+               [   0,    1,    0, 2000]], dtype=int16)
+
+    Oversized b-vectors are normalized and the b-value is scaled accordingly::
+
+        >>> format_gradients([[2.0, 0.0, 0.0, 1000]])
+        array([[1.e+00, 0.e+00, 0.e+00, 2.e+03]])
+
+    Near-zero b-vectors are suppressed to treat them as b0 measurements::
+
+        >>> format_gradients([[1e-9, 0.0, 0.0, 1200], [1.0, 0.0, 0.0, 1000]])
+        array([[   0,    0,    0,    0],
+               [   1,    0,    0, 1000]], dtype=int16)
+
+    Normaliztion can be skipped::
+
+        >>> format_gradients([[2.0, 0.0, 0.0, 1000]], skip_normalization=True)
+        array([[   2,    0,    0, 1000]], dtype=int16)
+
+    Integer-like inputs are preserved when no rescaling is needed::
+
+        >>> import numpy as np
+        >>> format_gradients(np.array([[0, 0, 0, 0], [0, 0, 1, 1000]], dtype=np.int16)).dtype
+        dtype('int16')
+
+    Passing ``None`` raises the absence error::
+
+        >>> format_gradients(None)
+        Traceback (most recent call last):
+        ...
+        ValueError: No gradient table was provided.
+
+    Gradient tables must always have two dimensions::
+
+        >>> format_gradients([0, 1, 0, 1000])
+        Traceback (most recent call last):
+        ...
+        ValueError: Gradient table must be a 2D array
+
+    Gradient tables must have a regular shape::
+
+        >>> format_gradients([[1, 2], [3, 4, 5]])
+        Traceback (most recent call last):
+        ...
+        TypeError: Gradient table must be a numeric homogeneous array-like object
+
+    Gradient tables must always have two dimensions::
+
+        >>> format_gradients([0, 1, 0, 1000])
+        Traceback (most recent call last):
+        ...
+        ValueError: Gradient table must be a 2D array
+
+    """
+
+    if value is None:
+        raise ValueError(GRADIENT_ABSENCE_ERROR_MSG)
+
+    try:
+        formatted = np.asarray(value, dtype=float)
+    except (TypeError, ValueError) as exc:
+        # Conversion failed (e.g. nested ragged objects, non-numeric)
+        raise TypeError(GRADIENT_OBJECT_ERROR_MSG) from exc
+
+    if formatted.ndim != 2:
+        raise ValueError(GRADIENT_NDIM_ERROR_MSG)
+
+    # If the numeric values are all integers, preserve integer dtype
+    if np.allclose(formatted, rounded := np.rint(formatted)):
+        formatted = rounded.astype(
+            formatted.dtype if np.issubdtype(formatted.dtype, np.integer) else np.int16
+        )
```

Comment:
Although it was myself who introduced a version of this to make doctests pass,
https://github.com/nipreps/nifreeze/pull/327/commits/064413eb3fb9e8deaa8da7190265f2ee502ffc41#diff-774e30bff24e702cda5516cbc6bbbcae1af21b710e2b0ad22cc15741fbed9afeR178
I later realized that since we will be normalizing gradients, gradients will anyway be returned as floats. 
Also, it did not make much sense to do this just to make doctests pass. In https://github.com/nipreps/nifreeze/pull/327/commits/7933b75e4f3c7623572b1fed61e2c9740d59e3b0#diff-774e30bff24e702cda5516cbc6bbbcae1af21b710e2b0ad22cc15741fbed9afeR119, I ended up by setting the NumPy formatter in an explicit way to make test pass and removed the block. Hope this makes sense.